pkg/cache/groupcache.go

```diff
@@ -0,0 +1,283 @@
+func NewGroupcacheWithConfig(logger log.Logger, reg prometheus.Registerer, conf GroupcacheConfig, basepath string, r *route.Router, bucket objstore.Bucket) (*Groupcache, error) {
+	dnsGroupcacheProvider := dns.NewProvider(
+		logger,
+		extprom.WrapRegistererWithPrefix("thanos_store_groupcache_", reg),
+		dns.ResolverType(conf.DNSSDResolver),
+	)
+
+	httpProto := galaxyhttp.NewHTTPFetchProtocol(&galaxyhttp.HTTPOptions{
+		BasePath: basepath,
+	})
+	universe := galaxycache.NewUniverse(httpProto, conf.SelfURL)
+
+	ticker := time.NewTicker(conf.DNSInterval)
+
+	go func() {
+		for {
+			if err := dnsGroupcacheProvider.Resolve(context.Background(), conf.Peers); err != nil {
+				level.Error(logger).Log("msg", "failed to resolve addresses for groupcache", "err", err)
+			} else {
+				err := universe.Set(dnsGroupcacheProvider.Addresses()...)
+				if err != nil {
+					level.Error(logger).Log("msg", "failed to set peers for groupcache", "err", err)
+				}
+			}
+
+			<-ticker.C
+		}
+	}()
+
+	mux := http.NewServeMux()
+	galaxyhttp.RegisterHTTPHandler(universe, &galaxyhttp.HTTPOptions{
+		BasePath: basepath,
+	}, mux)
+	r.Get(basepath, mux.ServeHTTP)
+
+	galaxy := universe.NewGalaxy(conf.GroupcacheGroup, int64(conf.MaxSize), galaxycache.GetterFunc(
+		func(ctx context.Context, id string, dest galaxycache.Codec) error {
+			parsedData, err := cachekey.ParseBucketCacheKey(id)
+			if err != nil {
+				return err
+			}
+
+			switch parsedData.Verb {
+			case cachekey.AttributesVerb:
+				attrs, err := bucket.Attributes(ctx, parsedData.Name)
+				if err != nil {
+					return err
+				}
+
+				finalAttrs, err := json.Marshal(attrs)
+				if err != nil {
+					return err
+				}
+				err = dest.UnmarshalBinary(finalAttrs)
+				if err != nil {
+					return err
+				}
+			case cachekey.IterVerb:
+				// Not supported.
+
+				return nil
+			case cachekey.ContentVerb:
+				rc, err := bucket.Get(ctx, parsedData.Name)
+				if err != nil {
+					return err
+				}
+				defer runutil.CloseWithLogOnErr(logger, rc, "closing get")
+
+				b, err := ioutil.ReadAll(rc)
+				if err != nil {
+					return err
+				}
+
+				err = dest.UnmarshalBinary(b)
+				if err != nil {
+					return err
+				}
+			case cachekey.ExistsVerb:
+				exists, err := bucket.Exists(ctx, parsedData.Name)
+				if err != nil {
+					return err
+				}
+
+				err = dest.UnmarshalBinary([]byte(strconv.FormatBool(exists)))
+				if err != nil {
+					return err
+				}
+			case cachekey.SubrangeVerb:
+				rc, err := bucket.GetRange(ctx, parsedData.Name, parsedData.Start, parsedData.End-parsedData.Start)
+				if err != nil {
+					return err
+				}
+				defer runutil.CloseWithLogOnErr(logger, rc, "closing get_range")
+
+				b, err := ioutil.ReadAll(rc)
+				if err != nil {
+					return err
+				}
+
+				err = dest.UnmarshalBinary(b)
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+	))
+
+	// Register GalaxyCache stats.
```

Comment:
Maybe we can move this on top of the function itself. And maybe we can reword it a bit to be more detailed.